x/exp/keeper/exp_request.go

```diff
@@ -0,0 +1,371 @@
+}
+
+// calculate exp value by ibc asset .
+func (k ExpKeeper) calculateDaoTokenValue(ctx sdk.Context, amount sdk.Int) sdk.Dec {
+	daoTokenPrice, _ := k.calculateDaoTokenPrice(ctx)
+
+	return daoTokenPrice.MulInt(amount)
+}
+
+func (k ExpKeeper) SetBurnRequest(ctx sdk.Context, burnRequest types.BurnRequest) {
+	store := ctx.KVStore(k.storeKey)
+	bz := k.cdc.MustMarshal(&burnRequest)
+
+	accAddress, err := sdk.AccAddressFromBech32(burnRequest.Account)
+	if err != nil {
+		panic(err)
+	}
+	store.Set(types.GetBurnRequestAddressBytes(int(burnRequest.Status), accAddress), bz)
+}
+
+func (k ExpKeeper) GetBurnRequestByKey(ctx sdk.Context, key []byte) (types.BurnRequest, error) {
+	var burnRequest types.BurnRequest
+
+	store := ctx.KVStore(k.storeKey)
+	if !store.Has(key) {
+		return types.BurnRequest{}, sdkerrors.Wrapf(types.ErrInvalidKey, "burnRequest")
+	}
+
+	bz := store.Get(key)
+	err := k.cdc.Unmarshal(bz, &burnRequest)
+	if err != nil {
+		return types.BurnRequest{}, err
+	}
+
+	return burnRequest, nil
+}
+
+// GetAllBurnRequest returns all the burn request from store .
+func (keeper ExpKeeper) GetAllBurnRequests(ctx sdk.Context) (burnRequests types.BurnRequests) {
+	keeper.IterateBurnRequests(ctx, func(burnRequest types.BurnRequest) bool {
+		burnRequests = append(burnRequests, burnRequest)
+		return false
+	})
+	return
+}
+
+// IterateBurnRequest iterates over the all the BurnRequest and performs a callback function .
+func (k ExpKeeper) IterateBurnRequests(ctx sdk.Context, cb func(burnRequest types.BurnRequest) (stop bool)) {
+	store := ctx.KVStore(k.storeKey)
+
+	iterator := sdk.KVStorePrefixIterator(store, types.KeyBurnRequestList)
+	defer iterator.Close()
+	for ; iterator.Valid(); iterator.Next() {
+		var burnRequest types.BurnRequest
+		err := k.cdc.Unmarshal(iterator.Value(), &burnRequest)
+		if err != nil {
+			panic(err)
+		}
+
+		if cb(burnRequest) {
+			break
+		}
+	}
+}
+
+// not good logic need modify
+func (k ExpKeeper) RemoveBurnRequest(ctx sdk.Context, burnRequest types.BurnRequest) {
+	store := ctx.KVStore(k.storeKey)
+	accAddress, _ := sdk.AccAddressFromBech32(burnRequest.Account)
+	if store.Has(types.GetBurnRequestAddressBytes(int(burnRequest.Status), accAddress)) {
+		store.Delete(types.GetBurnRequestAddressBytes(int(burnRequest.Status), accAddress))
+	}
+}
+
+func (k ExpKeeper) GetBurnRequestsByStatus(ctx sdk.Context, status int) (burnRequests types.BurnRequests) {
+	k.IterateStatusBurnRequests(ctx, status, func(burnRequest types.BurnRequest) bool {
+		burnRequests = append(burnRequests, burnRequest)
+		return false
+	})
+	return
+}
+
+// not good logic need modify
+func (k ExpKeeper) GetBurnRequest(ctx sdk.Context, accAddress sdk.AccAddress) (types.BurnRequest, error) {
+	store := ctx.KVStore(k.storeKey)
+
+	if store.Has(types.GetBurnRequestAddressBytes(int(types.StatusCompleteRequest), accAddress)) {
+		return k.GetBurnRequestByKey(ctx, types.GetBurnRequestAddressBytes(int(types.StatusCompleteRequest), accAddress))
+	}
```

Comment:
a lot of duplicate code and double checking, should also use a variable like:
```go
ab := types.GetBurnRequestAddressBytes(int(types.StatusCompleteRequest), accAddress)
```

GetBurnRquestByKey ( line 130) also already as a store.Has() check, so line 129 is a duplicate of what 130 already does

Is minor but can def be improved to reduce undeeded store queries

Reply:
Oh yep, it's my bad logic.

Reply:
@Reecepbcups, i'll change logic in kvstore again for better performance